BaseTools/Source/Python/build/build.py

```diff
@@ -837,13 +838,20 @@ def StartAutoGen(self,mqueue, DataPipe,SkipAutoGen,PcdMaList,cqueue):
         try:
             if SkipAutoGen:
                 return True,0
+            if sys.platform == "win32":
+                SafeThreadNumber = self.ThreadNumber
+            else:
+                import resource
+                soft = resource.getrlimit(resource.RLIMIT_NOFILE)[0]
+                SafeThreadNumber = min(self.ThreadNumber, soft // 4)

```

Comment:
As well as my recommendation to stick to 3, and add note on empirical evidence to commit message, it would be good to include an EdkLogger message, actually I would suggest at .info (always show) level, since I believe this only occurs once per package build.

If a user has specified a high thread count it would be good for this to be clearly surfaced in the build log.

Something like "Thread count limited to %d to avoid file descriptor exhaustion", ofc only to show when actually limited to less than the specified (or default, but ofc that won't happen) value.

Reply:
Apologies, that should be "Workers limited to...", as I see the specified thread count value is used in other locations not affected by this change.